services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java

```diff
@@ -7824,6 +7826,38 @@ public Optional<PushStatusStoreWriter> getPushStatusStoreWriter() {
     return pushStatusStoreWriter;
   }
 
+  @Override
+  public void sendHeartbeatToSystemStore(String clusterName, String storeName, long heartbeatTimeStamp) {
+    VeniceSystemStoreType systemStoreType = VeniceSystemStoreType.getSystemStoreType(storeName);
+    String userStoreName = systemStoreType.extractRegularStoreName(storeName);
+    long currentTimestamp = System.currentTimeMillis();
+    if (VeniceSystemStoreType.DAVINCI_PUSH_STATUS_STORE.equals(systemStoreType)) {
+      getPushStatusStoreWriter().get().writeHeartbeat(userStoreName, currentTimestamp);
+    } else {
+      getMetaStoreWriter().writeHeartbeat(userStoreName, currentTimestamp);
+    }
+  }
+
+  @Override
+  public long getHeartbeatFromSystemStore(String clusterName, String systemStoreName) {
+    VeniceSystemStoreType systemStoreType = VeniceSystemStoreType.getSystemStoreType(systemStoreName);
+    String userStoreName = systemStoreType.extractRegularStoreName(systemStoreName);
+    try {
+      return RetryUtils.executeWithMaxRetriesAndFixedAttemptDuration(() -> {
+        long retrievedTimestamp;
+        if (systemStoreType == VeniceSystemStoreType.DAVINCI_PUSH_STATUS_STORE) {
+          retrievedTimestamp = getPushStatusStoreReader().get()
+              .getHeartbeat(userStoreName, PushStatusStoreUtils.CONTROLLER_HEARTBEAT_INSTANCE_NAME);
+        } else {
+          retrievedTimestamp = getMetaStoreReader().getHeartbeat(userStoreName);
+        }
+        return retrievedTimestamp;
+      }, 3, Duration.ofSeconds(1), Collections.singletonList(VeniceException.class));
+    } catch (VeniceException e) {
+      return 0;
```

Comment:
Should we return the error response to Parent?
For example, we should return `-1` when error happens, and Parent Controller shouldn't try to repair on error response, but log it and maybe emit a metric in Parent Controller.

Reply:
I think this comment relates to the below comment right? So maybe just return -1 and check it and emit the error metrics.